nwapp/utils/notifier.js

```diff
@@ -28,14 +40,112 @@ function playNotificationSound() {
   audio = null;
 }
 
-module.exports = function (spec) {
-  if (!settings.showNotifications) return;
-  playNotificationSound();
 
-  var notification = new window.Notification(spec.title, {
-    body: spec.message,
-    icon: getNotificationImageForUri(spec.link)
+// Cache lookup
+var urlFilePathCacheMap = new Map();
+var tempAvatarDirectory = path.join(os.tmpDir(), 'gitter-notification-avatars_');
+log.info('Saving notification avatars at: ' + tempAvatarDirectory);
+var ensureTempAvatarDirectory = ensureDir(tempAvatarDirectory)
+  .then(function() {
+    return tempAvatarDirectory;
   });
+events.on('app:quit', function() {
+  log.info('App is quiting, cleanup avatars');
+  rimraf(tempAvatarDirectory)
+    .then(function() {
+      log.info('Finished avatar cleanup');
+    });
+});
+function getFileAtUrl(url) {
+  return ensureTempAvatarDirectory.then(function() {
+      var cachedFilePath = urlFilePathCacheMap.get(url);
+      if(!cachedFilePath) {
+        throw new Error('Avatar is not in the cache');
+      }
+      return cachedFilePath;
+    })
+    .catch(function(cachedFilePath) {
```

Comment:
Using try/catch for a cache miss is overkill as you're going to be missing the cache pretty often. Put the icon downloader in its own function that returns a promise so you can use it like this:

```js
function getFileUrlForHttpUrl(url) {
  return ensureCacheReady.then(function() {
    return cache.get(url) || downloadAndCache(url);
  })
  .then(function(filePath) {
    return 'file://' + filePath;
  });
}
```
(I've also simplified some of the function names)